### PR TITLE
Fix armhf build support

### PR DESCRIPTION
**Pull request by nuclearsandwich**
This adds more complete armhf support using the armhf images built for the ROS build farm on the native CI linux-arm workers, which support armv7 (armhf) and armv8 (aarch64) natively but are running as armv8 platforms.

Connects to ros2/ros2#721

### CODE REVIEW

File: linux_docker_resources/Dockerfile [hunk 1]
```diff
@@ -109,7 +109,7 @@ RUN if test ${UBUNTU_DISTRO} != xenial; then apt-get update && apt-get install -
 RUN if test ${UBUNTU_DISTRO} != xenial; then apt-get update && apt-get install --no-install-recommends -y libpoco-dev; fi
 
 # Install build dependencies for rviz et al.
-RUN apt-get update && apt-get install --no-install-recommends -y libassimp-dev libcurl4-openssl-dev libfreetype6-dev libgles2-mesa-dev libglu1-mesa libqt5core5a libqt5gui5 libqt5opengl5 libqt5widgets5 libxaw7-dev libxrandr-dev qtbase5-dev
+RUN apt-get update && apt-get install --no-install-recommends -y libassimp-dev libcurl4-openssl-dev libfreetype6-dev libgles2-mesa-dev libglu1-mesa-dev libqt5core5a libqt5gui5 libqt5opengl5 libqt5widgets5 libxaw7-dev libxrandr-dev qtbase5-dev
```
Comment (dirk-thomas) [inline linux_docker_resources/Dockerfile]: How is this change related to `armhf` support?

File: job_templates/ci_job.xml.em [hunk 2]
```diff
@@ -191,9 +191,12 @@ echo "Using args: $CI_ARGS"
 echo "# END SECTION"
 
 @[  if os_name in ['linux', 'linux-aarch64', 'linux-armhf', 'linux-centos']]@
-@[    if os_name in ['linux', 'linux-aarch64', 'linux-armhf']]@
+@[    if os_name in ['linux', 'linux-aarch64']]@
 sed -i "s+^FROM.*$+FROM ubuntu:$CI_UBUNTU_DISTRO+" linux_docker_resources/Dockerfile
 export DOCKER_BUILD_ARGS="${DOCKER_BUILD_ARGS} --build-arg UBUNTU_DISTRO=$CI_UBUNTU_DISTRO --build-arg ROS1_DISTRO=$CI_ROS1_DISTRO"
+@[    elif os_name == 'linux-armhf']@
+sed -i "s+^FROM.*$+FROM osrf/ubuntu_armhf:$CI_UBUNTU_DISTRO+" linux_docker_resources/Dockerfile
+export DOCKER_BUILD_ARGS="${DOCKER_BUILD_ARGS} --build-arg UBUNTU_DISTRO=$CI_UBUNTU_DISTRO --build-arg ROS1_DISTRO=$CI_ROS1_DISTRO"
```
Comment (dirk-thomas) [inline job_templates/ci_job.xml.em]: Nitpick: how about not duplicating this line?

**Comment by nuclearsandwich** [comment]
The test_ci_linux-armhf job has been deployed using this branch.

Most recent full build: [Build Status](https://ci.ros2.org/job/test_ci_linux-armhf/6/)
There are required packages not installed on armhf which get installed in our other environments. I've made some changes to improve the situation but more are missing libraries that need to be addressed.

Most recent build passing: [Build Status](https://ci.ros2.org/job/test_ci_linux-armhf/8/) in order to test the armhf opensplice deb I built as much as ROS 2 itself I triggered a build up to test_communication and test_rclcpp with Fast-RTPS and Opensplice. This build has some type size warnings and a glut of GCC ABI change warnings that @clalancette and I haven't quite gotten to the bottom of yet. They don't appear on other platforms because the change affects only armel and armhf targets.

File: linux_docker_resources/Dockerfile (hunk 1 above)
Comment (nuclearsandwich) [reply linux_docker_resources/Dockerfile]: Before [Build Status](https://ci.ros2.org/view/All/job/test_ci_linux-armhf/4/consoleFull#console-section-299)

After [Build Status](https://ci.ros2.org/job/test_ci_linux-armhf/5/)

I haven't taken the time to figure out what's pulling in the deb package or providing that header for our other platforms.

Comment (sloretz) [reply linux_docker_resources/Dockerfile]:
```
07:49:55 In file included from /home/jenkins-agent/workspace/test_ci_linux-armhf/ws/build/rviz_ogre_vendor/ogre-master-ca665a6-prefix/src/ogre-master-ca665a6/RenderSystems/GL/include/OgreGLPrerequisites.h:63:0,
07:49:55                  from /home/jenkins-agent/workspace/test_ci_linux-armhf/ws/build/rviz_ogre_vendor/ogre-master-ca665a6-prefix/src/ogre-master-ca665a6/RenderSystems/GL/src/GLSL/include/OgreGLSLExtSupport.h:34,
07:49:55                  from /home/jenkins-agent/workspace/test_ci_linux-armhf/ws/build/rviz_ogre_vendor/ogre-master-ca665a6-prefix/src/ogre-master-ca665a6/RenderSystems/GL/src/GLSL/src/OgreGLSLExtSupport.cpp:31:
07:49:55 /home/jenkins-agent/workspace/test_ci_linux-armhf/ws/build/rviz_ogre_vendor/ogre-master-ca665a6-prefix/src/ogre-master-ca665a6/RenderSystems/GL/include/GL/glew.h:1180:14: fatal error: GL/glu.h: No such file or directory
07:49:55  #    include <GL/glu.h>
07:49:55               ^~~~~~~~~~
```

Comment (sloretz) [reply linux_docker_resources/Dockerfile]: 1. [ogre_add_library_to_folder](https://github.com/OGRECave/ogre/blob/ca665a69a5ca47608ee0ad304b2596ff866616da/CMake/Utils/OgreAddTargets.cmake#L105)
1. [CMakeLists.txt](https://github.com/OGRECave/ogre/blob/ca665a69a5ca47608ee0ad304b2596ff866616da/RenderSystems/GL/CMakeLists.txt#L71)
1. [OgreGLSLExtSupport.cpp:31](https://github.com/OGRECave/ogre/blob/ca665a69a5ca47608ee0ad304b2596ff866616da/RenderSystems/GL/src/GLSL/src/OgreGLSLExtSupport.cpp#L31)
1. [OgreGLSLExtSupport.h:34](https://github.com/OGRECave/ogre/blob/ca665a69a5ca47608ee0ad304b2596ff866616da/RenderSystems/GL/src/GLSL/include/OgreGLSLExtSupport.h#L34)
1. [OgreGLPrerequisites.h:63](https://github.com/OGRECave/ogre/blob/ca665a69a5ca47608ee0ad304b2596ff866616da/RenderSystems/GL/include/OgreGLPrerequisites.h#L62-L64)
1. [GL/glew.h:1180](https://github.com/OGRECave/ogre/blob/ca665a69a5ca47608ee0ad304b2596ff866616da/RenderSystems/GL/include/GL/glew.h#L1175-L1182)

:man_shrugging:

Comment (cottsay) [reply linux_docker_resources/Dockerfile]: It should have been pulled in as a dependency of `qtbase5-dev` or `libsdl1.2-dev`. From a Bionic ARM instance:
```
$ apt-cache rdepends libglu1-mesa-dev
...
 |libsdl1.2-dev
...
 |qtbase5-dev
...
```

...the logs show both of those being installed, but the pipe indicates there are other options to satisfy that dependency. But the other option (libglu-dev) is a virtual package provided by libglu1-mesa-dev, so I have no idea how they were installed without libglu1-mesa-dev present as well.

File: job_templates/ci_job.xml.em (hunk 2 above)
Comment (cottsay) [reply job_templates/ci_job.xml.em]: Fixed.

File: job_templates/ci_job.xml.em [hunk 3]
```diff
@@ -192,7 +192,11 @@ echo "# END SECTION"
 
 @[  if os_name in ['linux', 'linux-aarch64', 'linux-armhf', 'linux-centos']]@
 @[    if os_name in ['linux', 'linux-aarch64', 'linux-armhf']]@
+@[      if os_name in ['linux-armhf']]@
```
Comment (jacobperron) [inline job_templates/ci_job.xml.em]: nitpick: could also be `'linux-armhf' == os_name`

File: linux_docker_resources/Dockerfile [hunk 4]
```diff
@@ -69,9 +69,9 @@ RUN if test ${COMPILE_WITH_CLANG} = true; then apt-get update && apt-get install
 RUN apt-get update && apt-get install --no-install-recommends -y gcovr
 
 # Install the OpenSplice binary from the OSRF repositories.
-RUN if test \( ${PLATFORM} = arm -o ${PLATFORM} = x86 \) ; then apt-get update && apt-get install --no-install-recommends -y libopensplice69=6.9.190403+osrf1-1~$UBUNTU_DISTRO; fi
+RUN if test \( ${PLATFORM} = aarch64 -o ${PLATFORM} = armhf -o ${PLATFORM} = x86 \) ; then apt-get update && apt-get install --no-install-recommends -y libopensplice69=6.9.190403+osrf1-1~$UBUNTU_DISTRO; fi
 # Update default domain id.
-RUN if test \( ${PLATFORM} = arm -o ${PLATFORM} = x86 \) ; then sed -i "s/<Id>0<\/Id>/<Id>108<\/Id>/" /usr/etc/opensplice/config/ospl.xml; fi
+RUN if test \( ${PLATFORM} = aarch64 -o ${PLATFORM} = x86 \) ; then sed -i "s/<Id>0<\/Id>/<Id>108<\/Id>/" /usr/etc/opensplice/config/ospl.xml; fi
```
Comment (sloretz) [inline linux_docker_resources/Dockerfile]: IIUC the change above this one installs opensplice if the platform is `aarch64`, `armhf`, or `x86`. Should this line be modified to change `ospl.xml` when the platform is `armhf` too?

Comment (cottsay) [reply linux_docker_resources/Dockerfile]: I don't see why not. What platform do we build besides `aarch64`, `armhf` and `x86`?

Comment (sloretz) [reply linux_docker_resources/Dockerfile]: I can't think of any. I do see job names on https://ci.ros2.org/ of `ci_linux-aarch64` and `ci_linux-centos`. Are our scripts treating CentOS as a different platform?

Comment (cottsay) [reply linux_docker_resources/Dockerfile]: CentOS has never used this Dockerfile. Even if it did, I think `PLATFORM` would be set to `x86` in this case.

Comment (cottsay) [reply linux_docker_resources/Dockerfile]: I can't find any mention of any PLATFORM value other than `aarch64`, `armhf` or `x86`. I think this is just a remnant of when we didn't install OpenSplice on one or more ARM platforms. I'll remove it, and trigger CI for all platforms.

**Comment by cottsay** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/7373/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/3531/)
* Linux-armhf [Build Status](http://ci.ros2.org/view/All/job/test_ci_linux-armhf/24/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/6048/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/7201/)

**Comment by sloretz** [comment]
CI LGTM. Windows CI has the same failures as the nightly: https://ci.ros2.org/view/nightly/job/nightly_win_rel/1222/#showFailuresLink

**Comment by cottsay** [comment]
Deployment complete :heavy_check_mark:

**Comment by emersonknapp** [comment]
Thanks @cottsay for the followup on this!

**Comment by emersonknapp** [comment]
@cottsay deployed, does that mean the environment should be in https://ci.ros2.org/view/All/job/ci_linux-armhf/ ? I am still seeing `fatal error: GL/glu.h: No such file or directory` when trying that build - https://ci.ros2.org/view/All/job/ci_linux-armhf/6/console
